vm.go

```diff
@@ -536,79 +556,174 @@ func (v *VM) run() {
-			v.ip += 2
+			var (
+				numArgs           = int(v.curInsts[v.ip+1])
+				hasVarArgs        = int(v.curInsts[v.ip+2])
+				numKws            = int(v.curInsts[v.ip+3])
+				hasVarKw          = int(v.curInsts[v.ip+4])
+				args              []Object
+				kwargs, varKwargs map[string]Object
+				hasKws            int
+			)
+
+			v.ip += 4
+			if numKws > 0 {
+				hasKws = 1
+			}
+
+			start := v.sp - numArgs - hasVarArgs - hasKws - hasVarKw
+			value := v.stack[start-1]
 
-			value := v.stack[v.sp-1-numArgs]
 			if !value.CanCall() {
 				v.err = fmt.Errorf("not callable: %s", value.TypeName())
 				return
 			}
 
-			if spread == 1 {
-				v.sp--
-				switch arr := v.stack[v.sp].(type) {
+			for i := 0; i < numArgs; i++ {
+				args = append(args, v.stack[start+i])
+			}
+
+			if hasVarArgs == 1 {
+				switch arr := v.stack[start+numArgs].(type) {
 				case *Array:
-					for _, item := range arr.Value {
-						v.stack[v.sp] = item
-						v.sp++
-					}
-					numArgs += len(arr.Value) - 1
+					newArgs := make([]Object, numArgs+len(arr.Value))
+					copy(newArgs, args)
+					copy(newArgs[numArgs:], arr.Value)
+					args = newArgs
 				case *ImmutableArray:
-					for _, item := range arr.Value {
-						v.stack[v.sp] = item
-						v.sp++
-					}
-					numArgs += len(arr.Value) - 1
+					newArgs := make([]Object, numArgs+len(arr.Value))
+					copy(newArgs, args)
+					copy(newArgs[numArgs:], arr.Value)
+					args = newArgs
 				default:
 					v.err = fmt.Errorf("not an array: %s", arr.TypeName())
 					return
 				}
 			}
 
-			if callee, ok := value.(*CompiledFunction); ok {
-				if callee.VarArgs {
-					// if the closure is variadic,
-					// roll up all variadic parameters into an array
-					realArgs := callee.NumParameters - 1
-					varArgs := numArgs - realArgs
-					if varArgs >= 0 {
-						numArgs = realArgs + 1
-						args := make([]Object, varArgs)
-						spStart := v.sp - varArgs
-						for i := spStart; i < v.sp; i++ {
-							args[i-spStart] = v.stack[i]
+			if hasKws == 1 {
+				kwargs = v.stack[start+numArgs+hasVarArgs].(*Map).Value
+			}
+
+			if hasVarKw == 1 {
+				if kwargs == nil {
+					switch t := v.stack[start+numArgs+hasVarArgs+hasKws].(type) {
+					case *Map:
+						kwargs = t.Value
+					case *ImmutableMap:
+						kwargs = t.Value
+					}
+				} else {
+					switch t := v.stack[start+numArgs+hasVarArgs+hasKws].(type) {
+					case *Map:
+						for k, v := range t.Value {
+							kwargs[k] = v
+						}
+					case *ImmutableMap:
+						for k, v := range t.Value {
+							kwargs[k] = v
 						}
-						v.stack[spStart] = &Array{Value: args}
-						v.sp = spStart + 1
 					}
 				}
-				if numArgs != callee.NumParameters {
-					if callee.VarArgs {
+			}
+
+			if callee, ok := value.(*CompiledFunction); ok {
+				calleeData := &ImmutableMap{
+					Value: map[string]Object{
+						"fn":     callee,
+						"args":   &ImmutableArray{Value: args},
+						"kwargs": &ImmutableMap{Value: kwargs},
+					},
+				}
```

Comment:
Also try to see if you can avoid recreating this and update the existing one instead